acceptance/managers/container_manager.go

```diff
@@ -0,0 +1,278 @@
+// +build acceptance
+
+package managers
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"strings"
+	"testing"
+	"time"
+
+	dockertypes "github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/client"
+	"github.com/docker/docker/pkg/stdcopy"
+	"github.com/docker/go-connections/nat"
+
+	"github.com/buildpacks/pack/acceptance/assertions"
+	h "github.com/buildpacks/pack/testhelpers"
+)
+
+type TestContainerManager struct {
+	testObject *testing.T
+	assert     assertions.AssertionManager
+	dockerCli  *client.Client
+}
+
+//nolint:whitespace // A leading line of whitespace is left after a method declaration with multi-line arguments
+func NewTestContainerManager(
+	t *testing.T,
+	assert assertions.AssertionManager,
+	dockerCli *client.Client,
+) TestContainerManager {
+
+	return TestContainerManager{
+		testObject: t,
+		assert:     assert,
+		dockerCli:  dockerCli,
+	}
+}
+
+func (c TestContainerManager) HostOS() string {
+	c.testObject.Helper()
+
+	daemonInfo, err := c.dockerCli.Info(context.Background())
+	c.assert.Nil(err)
+
+	return daemonInfo.OSType
+}
+
+func (c TestContainerManager) CleanupTaskForImageByName(imageName string) func() {
```

Comment:
I found the pattern of calling this and then `defer`ing the returned func a bit unnecessary; it could have just defered a CleanupImageByName func which did all of this.

Reply:
This is tricky, and I think I may have introduced some leakage through inconsistent application of `defer`ing the returned method.

What I _think_ is happening, is that some of our `pack` invocations will create copies of images and in doing so rename the existing image. This task effectively transforms that name into an ID which is then encapsulated in the `defer`red function and _should_ ensure we clean up the original images as well as the copied ones.

I'm not certain that's what's happening though, but this approach does mirror what's happening in our existing tests.

I think I have a couple of areas to investigate out of this:
* Are my assumptions around image renaming accurate, or have I simply inherited existing poor patterns
* If this is necessary, how can we accurately communicate the intent of returning a deferrable function rather than the more obvious approach of deferring the `CleanupImageByName`
* Again, if this is necessary, how can ensure that people adding new tests don't follow the *obvious* and on the face of it correct path (which I have done in this PR many times)
* Do we need to bake in stronger guarantees around our cleanup into our test suite?

Reply:
> This is tricky, and I think I may have introduced some leakage

**With PR**:
![With PR](https://user-images.githubusercontent.com/7035673/85332481-06361580-b4a6-11ea-8e3e-20597846f59d.png)

**Pre PR**:
![Pre PR](https://user-images.githubusercontent.com/7035673/85332699-65942580-b4a6-11ea-8b5e-c9c5235c3cf2.png)

There definitely does seem to be some leakage from the tests, though there was some minor leakage from before the refactor as well (I would expect the `registry` to be cleaned up as well)